torch/csrc/dynamo/eval_frame.c

```diff
@@ -314,31 +396,228 @@ static void destroy_cache_entry(CacheEntry* e) {
+    return ((CacheEntryWrapper*)extra)->cache_entry;
+  }
+
+  DEBUG_CACHE("Getting cache entry for a nn module %p", nn_module);
+
+  // Case 4.
+  DEBUG_CHECK(PyDict_Check(extra));
+  DEBUG_NULL_CHECK(nn_module);
+
+  // TODO - the callback is set to NULL. Currently, there is a callback in
+  // CheckFnManager invalidate, which should track the garbage collection of nn
+  // modules. However, we can't fully rely on that one to call reset_code
+  // because when the callback is called, the referent is dead, so the weakrefs
+  // in the C cache will no longer hold equality with the weakref from python.
+  // Revisit this callback later, when we want to automatically free the cached
+  // guards/graphs on module garbage collection. This involves understanding how
+  // to write a Python callback function purely in C.
+  PyObject* nn_module_weakref = PyWeakref_NewRef(nn_module, NULL);
+
+  // If the cache is empty, return a null object.
+  PyObject* nn_module_to_cache_entry_map = extra;
+  if (PyDict_GetItem(nn_module_to_cache_entry_map, nn_module_weakref) == NULL) {
+    return NULL;
+  }
+
+  // Find the cache entry.
+  CacheEntryWrapper* cache_entry_wrapper = (CacheEntryWrapper*)PyDict_GetItem(nn_module_to_cache_entry_map, nn_module_weakref);
+  CacheEntry* a = cache_entry_wrapper->cache_entry;
+  return a;
+}
+
+inline static void set_cache_entry_on_code(PyCodeObject* code, CacheEntry* extra) {
   _PyCode_SetExtra((PyObject*)code, cache_entry_extra_index, extra);
 }
 
-inline static PyObject* get_frame_state(PyCodeObject* code) {
+inline static void set_cache_entry(PyCodeObject* code, CacheEntry* cache_entry, PyObject* nn_module) {
+  // The cache lives on the extra segment of code object. It can have one of the 4 possible values
+  // 1) NULL - First time accessed the extra segment.
+  // 2) SKIP_CODE - We will skip the original frame.
+  // 3) CacheEntryWrapper - This contains the cache_entry for the frame.
+  // 4) A dict from nn module to the CacheEntryWrapper.
+  // Look in get_cache_entry comments for more details.
+
+  // TODO(jansel): would it be faster to bypass this?
+
+  // Case 1 and 2 - Short circuit for SKIP_CODE.
+  if (cache_entry == NULL || cache_entry == SKIP_CODE) {
+    set_cache_entry_on_code(code, cache_entry);
+    return;
+  }
+  // Case 3
+
+  if (nn_module == NULL) {
+    CacheEntryWrapper *cache_entry_wrapper = (CacheEntryWrapper*) PyObject_CallObject((PyObject *) &CacheEntryWrapperType, NULL);
+    cache_entry_wrapper->cache_entry = cache_entry;
+    _PyCode_SetExtra((PyObject*)code, cache_entry_extra_index, (PyObject*) cache_entry_wrapper);
+    return;
+  }
+  DEBUG_CACHE("Setting cache entry for a nn module %p", nn_module);
+
+  // Case 4
+  PyObject* nn_module_to_cache_entry_map = NULL;
+  _PyCode_GetExtra((PyObject*)code, cache_entry_extra_index, (void*)&nn_module_to_cache_entry_map);
+
+  if (nn_module_to_cache_entry_map == NULL) {
+    nn_module_to_cache_entry_map = PyDict_New();
```

Comment:
Using a PyDict and not a WeakKeyDictionary means that even when weak references become dead, you'll continue to maintain the cache entry here, thereby leaking compiled code objects. This will be very bad if a nn Module is being created every iteration.

Reply:
Speaking of an nn Module being created every iteration, we may need some larger compile cache limit that helps avoid degenerate recompilation in this case.

Reply:
Thanks for pointing to WeakKeyDictionary. Let me read more about it.